### PR TITLE
Bring L1T (nearly) too l1t-tsg-v5 (81x version)

### PR DESCRIPTION
**Pull request by mulhearn**
This PR brings L1T very close to l1t-tsg-v5. It only differs by updates to Tau LUTs requested for inclusion by TSG (but not used yet by default).

We can integrate and merge this as is, and make a tiny update later when the externals are ready.

Unfortunately, due to updates to Provenance, the same branch cannot be used in both 80x and 81x.  As a result, I have to scrap PR #14091.  So this PR supersedes #14091, #14025, and #14026.

Now that we have removed the use of Provenance, there should be no difference with the 80x version of the PR (#14092), once we cherry pick the removal commit.

### CODE REVIEW

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
Those RelVals (10021, 10024) pass for me in CMSSW_8_1_X_2016-04-14-2300...

I can reproduce the AddOn failures.  It looks like maybe L1TCaloLayer1 is interacting poorly with 
recent changes from HCAL:
```
#5  0x00007fbc30339506 in CaloTPGTranscoderULUT::hcaletValue(HcalTrigTowerDetId const&, HcalTriggerPrimitiveSample const&) const () from /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-04-14-2300/lib/slc6_amd64_gcc530/libCalibCalorimetryCaloTPG.so
#6  0x00007fbbeede13e8 in L1TCaloLayer1FetchLUTs(edm::EventSetup const&, std::vector<std::vector<std::vector<unsigned int, std::allocator<unsigned int> >, std::allocator<std::vector<unsigned int, std::allocator<unsigned int> > > >, std::allocator<std::vector<std::vector<unsigned int, std::allocator<unsigned int> >, std::allocator<std::vector<unsigned int, std::allocator<unsigned int> > > > > >&, std::vector<std::vector<std::vector<unsigned int, std::allocator<unsigned int> >, std::allocator<std::vector<unsigned int, std::allocator<unsigned int> > > >, std::allocator<std::vector<std::vector<unsigned int, std::allocator<unsigned int> >, std::allocator<std::vector<unsigned int, std::allocator<unsigned int> > > > > >&, std::vector<std::vector<unsigned int, std::allocator<unsigned int> >, std::allocator<std::vector<unsigned int, std::allocator<unsigned int> > > >&, bool, bool, bool, bool, bool) () from /afs/cern.ch/work/m/mulhearn/trigger/v5_validate/CMSSW_8_1_X_2016-04-14-2300/lib/slc6_amd64_gcc530/libL1TriggerL1TCaloLayer1.so
```
Maybe this guy:
```
commit abd4163622272f195b1e1ee533acd02212fe6a28
Merge: a329cf2 691b3ac
Author: cmsbuild <cmsbuild@cern.ch>
Date:   Thu Apr 14 12:58:32 2016 +0200

    Merge pull request #13252 from matz-e/fix-calotpg-hcalet
    
    Fix the hcaletValue function.

```

**Comment by mulhearn** [comment]
```
10021.0_TenMuE_0_200+TenMuE_0_200_pythia8_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017 Step0-PASSED Step1-PASSED Step2-PASSED Step3-PASSED  - time date Sat Apr 16 03:03:52 2016-date Sat Apr 16 02:54:04 2016; exit: 0 0 0 0
10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017 Step0-PASSED Step1-PASSED Step2-PASSED Step3-PASSED  - time date Sat Apr 16 03:14:20 2016-date Sat Apr 16 02:54:06 2016; exit: 0 0 0 0
2 2 2 2 tests passed, 0 0 0 0 failed
```

**Comment by mulhearn** [comment]
I added some range protection which fixes the AddOn tests in my local tests.  We should still understand why this is going out of range however.  I see many instances of my new error:
```
%MSG
%MSG-e CaloTPGTranscoderULUT:  L1TCaloLayer1:simCaloStage2Layer1Digis@beginRun  16-Apr-2016 09:11:46 CEST Run: 256677
ERROR:  compET:  254
ERROR: itower:  4573
ERROR:  hcaluncomp_.size():  4176
ERROR:  hcaluncomp_[itower].size()]:  4176
```

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
+1

**Comment by mulhearn** [comment]
I just noticed that this changes the persistent data in CondFormats, which was not supposed to be the case.  I'll fix it.  This is supposed to be cosmetic change only (simply moving interpreter class outside of CondFormats).

**Comment by mulhearn** [comment]
I have a fix, which I'll push shortly, but I think its best if I remove the ALCA/DB changes entirely for now.  Takes a bit of updating.

**Comment by mulhearn** [comment]
Now this version leaves persistent data in CondFormats unchanged, so ALCA/DB signoff should be possible.  But if not, I'll have a version that factors off the CondFormats changes entirely by tonight.  Gotta go play with kids for a bit!

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
That takes DB off the required signoffs, but ALCA is still on due to our fix of CalibCalorimetry...

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
ALCA and DB signoff is not required anymore.  Now this is my kind of party!

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
Now the Layer1 checks the HCAL geometry and if use1x1=0 it very politely point outs that maybe using such a condition and Layer1 calo is not wise, and then disables HF, keeping everyone happy.

**Comment by mulhearn** [comment]
+1

**Comment by mulhearn** [comment]
@davidlange6 this one is ready to go!

**Comment by fwyzard** [comment]
> This PR brings L1T very close to l1t-tsg-v5. It only differs by updates to Tau LUTs requested for inclusion by TSG (but not used yet by default).

Interesting choice of words, given that there isn't any "l1t-tsg-v5" tag in the cms-l1t-offline repository.

**Comment by fwyzard** [comment]
not an easy feat... the commit history of this PR is messy, with changes being applied and then reverted, which should then be squashed out, and yet fails to give any information about the actual changes.
For example

7a6f3c3 L1T data formats updates for l1t-tsg-v5

is way less informative than

6640fd5 make tau vars methods const and store them
b6f0d4e Refactor BXVector::set() to replace erase/insert.

**Comment by mulhearn** [comment]
- Waiting on results of tests to define l1t-tsg-v5.  The plan was to get this PR ready and then have only tiny update, as explained above.
- I've already discussed the commit quashing procedure used for this PR elsewhere.  It is globally optimal in the current context, though there are certainly find sub-optimal features.  You found one!
- Note the reverted commit is actually a useful test, so I'm leaving it here.

**Comment by fwyzard** [comment]
I disagree, more or less strongly, on all three points. 

> Waiting on results of tests to define l1t-tsg-v5. The plan was to get this PR ready and then have only tiny update, as explained above.

This was fine a month ago. It is not fine now. We had a meeting on Thursday, and the conclusion was to cut `l1t-tsg-v5` with what we had at the end of the meeting - knowing full well that it would not be a "perfect" tag, possibly not even a "correct" tag, but that we *needed* a tag *then*, not two days from now.
My preference at the time would have been to go with `l1t-tsg-v5` equal to `l1t-integration-v36.0`.
Since then
  - `l1t-integration-v37.0` makes a physics-affecting change, and again this was integrated *before* people showed if it was actually improving things or not
  -  `l1t-integration-v37.1` adds completely orthogonal changes
  - `l1t-integration-v37.2` fixes a bug, but raises the question why do we have both `L1Trigger/L1TCalorimeter/python/caloStage2Params_cfi.py` and `L1Trigger/L1TCalorimeter/python/caloStage2Params_2016_v2.0_cfi.py`, and why are they out of sync ?

So now we are again in a situation were we do not know what are the L1 performance using the top of the integration branch.

> I've already discussed the commit quashing procedure used for this PR elsewhere. It is globally optimal in the current context, though there are certainly find sub-optimal features. You found one!

I believe that what you have done is not optimal, neither from the developers point of view, nor from the commit history that ends up in CMSSW.
If you want to go for a simple commit history, make a single commit with a message like "L1T updates more or less in sync with l1t-tsg-v5". 

> Note the reverted commit is actually a useful test, so I'm leaving it here.

As you write it is a useful *test* - but not something that should go in the final commit history.

**Comment by mulhearn** [comment]
You misunderstand.  It will be useful to have that commit included in a future PR, so I leave it as a revert.

**Comment by mulhearn** [comment]
There appears to be some confusion.  The integration branch is our private integration branch.

**Comment by fwyzard** [comment]
> You misunderstand. It will be useful to have that commit included in a future PR, so I leave it as a revert.

I understand, and I'm saying that it is the wrong approach.
Squash the commit/revert in this PR, and simply add the commit in a future PR.

> There appears to be some confusion. The integration branch is our private integration branch.

It stops being your private integration branch if you base on it the studies that are presented to the collaboration and used to produce the final L1 and HLT menus for data taking.

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
This is exactly equivalent to the previous version, but with the marginally improved history suggested by Andrea.

**Comment by fwyzard** [comment]
Well, mine was an example based on one of the commits, but the principle applies to the other as well.

Anyway, the rollbacks have mostly been squashed, so the history is at least cleaner.

Thanks,
.A

**Comment by mulhearn** [comment]
+1

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
This is bitwise equivalent to previous version in my tests... now the settings come from config instead of guessing from Provenance.

**Comment by mulhearn** [comment]
+1

File: L1Trigger/L1TCommon/plugins/L1TComparison.cc [hunk 1]
```diff
@@ -361,6 +361,23 @@ L1TComparison::endJob() {
   if (jetCheck_)  cout << "jet:   " << jetFails_ << "\n";
   if (sumCheck_)  cout << "sum:   " << sumFails_ << "\n";
   if (muonCheck_) cout << "muon:  " << muonFails_ << "\n";
+
+  int fail = 0;
```
Comment (davidlange6) [inline L1Trigger/L1TCommon/plugins/L1TComparison.cc]: @mulhearn - why not a bool?

File: L1Trigger/L1TMuon/interface/L1TMuonGlobalParamsHelper.h [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+  L1TMuonGlobalParamsHelper() { pnodes_.resize(NUM_GMTPARAMNODES); }
+  L1TMuonGlobalParamsHelper(const L1TMuonGlobalParams &);
+  ~L1TMuonGlobalParamsHelper() {}
+
+  // FW version
+  unsigned fwVersion() const { return pnodes_[FWVERSION].uparams_.size() > FWVERSION_IDX ? pnodes_[FWVERSION].uparams_[FWVERSION_IDX] : 0; }
+  void setFwVersion(unsigned fwVersion);
+
+  // Input disables
+  std::bitset<72> inputsToDisable() const { return inputFlags(INPUTS_TO_DISABLE); };
```
Comment (davidlange6) [inline L1Trigger/L1TMuon/interface/L1TMuonGlobalParamsHelper.h]: @mulhearn - lets think about how to replace all of these magic numbers

**Comment by davidlange6** [comment]
+1

**Comment by mulhearn** [comment]
Agreed.  I've appended it to our homework TODO list here:  https://github.com/cms-l1t-offline/cmssw/issues/220 where you can also see that we've been working on the other items in good faith!

**Pull request merged**